spring-cloud-gcp-pubsub/src/test/java/org/springframework/cloud/gcp/pubsub/core/PubSubTemplateTest.java

```diff
@@ -0,0 +1,129 @@
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.mockito.runners.MockitoJUnitRunner;
+
+import org.springframework.cloud.gcp.pubsub.support.PublisherFactory;
+import org.springframework.messaging.Message;
+import org.springframework.messaging.MessageHeaders;
+import org.springframework.messaging.converter.MessageConversionException;
+import org.springframework.messaging.converter.MessageConverter;
+import org.springframework.messaging.support.GenericMessage;
+import org.springframework.util.concurrent.ListenableFuture;
+
+import static org.junit.Assert.assertEquals;
+import static org.mockito.Matchers.isA;
+import static org.mockito.Mockito.when;
+
+/**
+ * @author João André Martins
+ */
+@RunWith(MockitoJUnitRunner.class)
+public class PubSubTemplateTest {
+
+	@Mock
+	private PublisherFactory mockPublisherFactory;
+	@Mock
+	private Publisher mockPublisher;
+
+	private PubSubTemplate pubSubTemplate;
+	private Message<String> siMessage;
+	private SettableApiFuture<String> settableApiFuture;
+
+	@Before
+	public void setUp() throws ExecutionException, InterruptedException {
+		this.pubSubTemplate = new PubSubTemplate(this.mockPublisherFactory);
+		when(this.mockPublisherFactory.getPublisher("testTopic")).thenReturn(this.mockPublisher);
+		this.settableApiFuture = SettableApiFuture.create();
+		when(this.mockPublisher.publish(isA(PubsubMessage.class)))
+				.thenReturn(this.settableApiFuture);
+
+		Map<String, Object> headers = new HashMap<>();
+		headers.put("header1", "value1");
+		headers.put("header2", 12345);
+
+		this.siMessage = new GenericMessage<>("testBody", headers);
+	}
+
+	@Test
+	public void testSend() throws ExecutionException, InterruptedException {
+		ListenableFuture<String> future = this.pubSubTemplate.send("testTopic", this.siMessage);
+		this.settableApiFuture.set("result");
+
+		assertEquals("result", future.get());
+	}
+
+	@Test(expected = MessageConversionException.class)
+	public void testSend_notPubsubMessage() {
+		this.pubSubTemplate.setMessageConverter(new MessageConverter() {
+			@Override
+			public Object fromMessage(Message<?> message, Class<?> targetClass) {
+				return "not PubsubMessage.class";
+			}
+
+			@Override
+			public Message<?> toMessage(Object payload, MessageHeaders headers) {
+				return null;
+			}
+		});
+
+		this.pubSubTemplate.send("testTopic", this.siMessage);
+	}
+
+	@Test(expected = PubSubException.class)
+	public void testSend_noPublisher() {
+		when(this.mockPublisherFactory.getPublisher("testTopic"))
+				.thenThrow(new PubSubException("couldn't create the publisher."));
+
+		this.pubSubTemplate.send("testTopic", this.siMessage);
+	}
+
+	@Test
+	public void testSend_onFailure() {
+		ListenableFuture<String> future = this.pubSubTemplate.send("testTopic", this.siMessage);
+		this.settableApiFuture.setException(new Exception("future failed."));
+
+		try {
+			future.get();
+			TestCase.fail("Test should fail.");
```

Comment:
Please, don't use `junit.framework` pacakge. Don't remember why, but `fail()` can simply be imported statically from the `Assert` class.